tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusAssembleToolsTask.groovy

```diff
@@ -0,0 +1,209 @@
+import org.gradle.api.file.RegularFileProperty
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.ListProperty
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.Nested
+import org.gradle.api.tasks.OutputFile
+import org.gradle.api.tasks.TaskAction
+
+import javax.inject.Inject
+
+import groovy.transform.Immutable
+
+/**
+ * A task that fetches a prebuilt `nimbus-fml` binary for the current platform.
+ *
+ * Prebuilt binaries for all platforms are packaged into ZIP archives, and
+ * published to sources like `archive.mozilla.org` (for releases) or
+ * TaskCluster (for nightly builds).
+ *
+ * This task takes a variable number of `@Input`s: a list of archive sources,
+ * and a list of glob patterns to find the binary for the current platform
+ * in the archive. It produces three `@OutputFile`s: the fetched archive and
+ * hash file, and the unzipped binary.
+ *
+ * The unzipped binary is used as an optional input for each
+ * `NimbusFmlCommandTask`.
+ */
+@CacheableTask
+abstract class NimbusAssembleToolsTask extends DefaultTask {
+    @Inject
+    abstract ArchiveOperations getArchiveOperations()
+
+    @Nested
+    abstract FetchSpec getFetchSpec()
+
+    @Nested
+    abstract UnzipSpec getUnzipSpec()
+
+    /** The location of the fetched ZIP archive. */
+    @OutputFile
+    abstract RegularFileProperty getArchiveFile()
+
+    /**
+     * The location of the fetched hash file, which contains the
+     * archive's checksum.
+     */
+    @OutputFile
+    abstract RegularFileProperty getHashFile()
+
+    /** The location of the unzipped binary. */
+    @OutputFile
+    abstract RegularFileProperty getFmlBinary()
+
+    /**
+     * Configures the task to download the archive.
+     *
+     * @param action The configuration action.
+     */
+    void fetch(Action<FetchSpec> action) {
+        action.execute(fetchSpec)
+    }
+
+    /**
+     * Configures the task to extract the binary from the archive.
+     *
+     * @param action The configuration action.
+     */
+    void unzip(Action<UnzipSpec> action) {
+        action.execute(unzipSpec)
+    }
+
+    @TaskAction
+    void assembleTools() {
+        def sources = [fetchSpec, *fetchSpec.alternativeSources.get()].collect {
+            new Source(new URI(it.archive.get()), new URI(it.hash.get()))
+        }
+
+        def successfulSource = sources.find { it.trySaveArchiveTo(archiveFile.get().asFile) }
+        if (successfulSource == null) {
+            throw new GradleException("Couldn't fetch archive from any of: ${sources*.archiveURI.collect { "`$it`" }.join(', ')}")
+        }
+
+        // We get the checksum, although don't do anything with it yet;
+        // Checking it here would be able to detect if the zip file was tampered with
+        // in transit between here and the server.
+        // It won't detect compromise of the CI server.
```

Comment:
For releases on `archive.mozilla.org`, we could validate the checksum by embedding the hexdigest for each release into this source file, then piping the ZIP file through a `DigestInputStream` as it's downloading, and verifying that the digests match after. That would be super tedious to do manually for nightlies, though.

I kept the existing logic to download the hash file to minimize churn, but I agree with the sentiment in the comment; it doesn't feel like it buys us a ton now.